[nnc] Use given kernel function name while emitting code (#66216)

navahgar opened the pull request:
Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/66216

Test Plan: Imported from OSS

Differential Revision: D32145958

Pulled By: navahgar

ZolotukhinM reviewed (approved):
LGTM! Is it possible to add a test where we actually set some custom functions name too?

navahgar commented:
> LGTM! Is it possible to add a test where we actually set some custom functions name too?

Updated the test to check for custom function name.

navahgar committed:
[nnc] Use given kernel function name while emitting code (#67781)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/67781

Update `LLVMCodeGen` in NNC to use the given kernel function name while emitting code.

This was earlier committed as D31445799 (https://github.com/pytorch/pytorch/commit/c30dc52739c083d8b10f36ed3e84f765f349fe9b) and got reverted as part of a stack of diffs that included a cache for `PyTorchLLVMJIT`, which was the likely culprit.

Test Plan:
```
buck test mode/opt //caffe2/test/cpp/tensorexpr:tensorexpr -- --exact 'caffe2/test/cpp/tensorexpr:tensorexpr - LLVM.CodeGenKernelFuncName'
```

Reviewed By: ZolotukhinM, bdhirsh

Differential Revision: D32145958

fbshipit-source-id: d066efa46b66c9bf6872b5f1efef8d13632d2e47
